cygnus-common/src/main/java/com/telefonica/iot/cygnus/backends/sql/SQLQueryUtils.java

```diff
@@ -140,19 +132,114 @@ protected static StringBuffer sqlUpsertQuery(LinkedHashMap<String, ArrayList<Jso
+                                                        LinkedHashMap<String, ArrayList<JsonElement>> lastData,
+                                                        String tableName,
+                                                        String tableSuffix,
+                                                        String uniqueKey,
+                                                        String timestampKey,
+                                                        String timestampFormat,
+                                                        String sqlInstance,
+                                                        String destination) {
+
+        StringBuffer updateSet = new StringBuffer();
+        StringBuffer query = new StringBuffer();
+        StringBuffer dateKeyUpdate = new StringBuffer();
+        boolean first = true;
+
+        for (String key : lastData.keySet()) {
+            if (!key.equals(timestampKey)) {
+                if (!key.equals(uniqueKey) && first) {
+                    first = false;
+                } else if (!key.equals(uniqueKey)) {
+                    updateSet.append(", ");
+                }
+                if (!key.equals(uniqueKey)) {
+                    updateSet.append(mySQLUpdateRecordQuery(key, uniqueKey, timestampKey, timestampFormat));
+                }
+            } else {
+                dateKeyUpdate.append(mySQLUpdateRecordQuery(key, uniqueKey, timestampKey, timestampFormat));
+            }
+        }
+        // The key that corresponds to the timestampKey must be updated at the end, if it is updated before any other key.
+        // The subsecuent timestamp validations will be allways false, because the date would already be updated.
+        if (first) {
+            updateSet.append(dateKeyUpdate);
+        } else {
+            updateSet.append(", ").append(dateKeyUpdate);
         }
+
+        StringBuffer insertQuery = sqlInsertQuery(lastData,
+                tableName.concat(tableSuffix),
+                sqlInstance,
+                destination);
+
+        query.append(insertQuery).
+                append("ON DUPLICATE KEY ").
+                append("UPDATE ").append(updateSet);
+
         LOGGER.debug("[NGSISQLUtils.sqlUpsertQuery] Preparing Upsert query: " + query.toString());
         return query;
     }
 
+    /**
+     * Creates a update statement for an upsert query
+     *
+     * @param key     the table suffix
+     * @param uniqueKey       the unique key
+     * @param timestampKey    the timestamp key
+     * @param timestampFormat the timestamp format
+     * @return the string buffer like the following one
+     * recvTime=IF((entityId=VALUES(entityId)) AND (STR_TO_DATE(recvTime, '%Y-%m-%d %H:%i:%s.%f') < (STR_TO_DATE(VALUES(recvTime), '%Y-%m-%d %H:%i:%s.%f'))), VALUES(recvTime), recvTime)
+     */
+
+    protected static StringBuffer mySQLUpdateRecordQuery(String key,
+                                                       String uniqueKey,
```

Comment:
Not sure if indentation is right

Reply:
Fixed in [e0aa884](https://github.com/telefonicaid/fiware-cygnus/pull/1944/commits/e0aa8847b5596ace1a4bf45340f279f7aa8e90e0)